File: Codility/9_maximum_slice_problem/max_profit.py
# https://codility.com/programmers/task/max_profit/

def solution(A):
    max_profit_ending = 0
    max_profit = 0
    for i in xrange(1, len(A)):
        change = A[i] - A[i - 1]
        max_profit_ending = max(0, max_profit_ending + change)
        max_profit = max(max_profit, max_profit_ending)
    
    return max_profit

File: Codility/9_maximum_slice_problem/max_profit.cpp
// https://codility.com/programmers/task/max_profit/

int solution(vector<int> &A) {
    int maxProfit = 0;
    int maxProfitEnding = 0;
    for (size_t i = 1; i < A.size(); ++i)
    {
        const int change = A[i] - A[i - 1];
        maxProfitEnding = max(0, maxProfitEnding + change);
        maxProfit = max(maxProfit, maxProfitEnding);
    }
    return maxProfit;
}